beacon-chain/sync/data_columns_sampling.go

```diff
@@ -273,132 +294,188 @@ func (s *Service) incrementalDAS(
-			"currentSampleCount":  oldExtendedSampleCount,
-			"nextSampleCount":     extendedSampleCount,
+			"missingColumnsCount": allowedFailures,
+			"currentSampleIndex":  oldExtendedSampleCount,
+			"nextSampleIndex":     extendedSampleCount,
 		}).Debug("Some columns are still missing after sampling this round.")
 	}
 }
 
-// DataColumnSamplingRoutine runs incremental DAS on block when received.
-func (s *Service) DataColumnSamplingRoutine(ctx context.Context) {
-	// Get the custody subnets count.
-	custodySubnetsCount := peerdas.CustodySubnetCount()
-
-	// Create a subscription to the state feed.
-	stateChannel := make(chan *feed.Event, 1)
-	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateChannel)
+func (d *dataColumnSampler1D) sampleDataColumns(
+	ctx context.Context,
+	root [fieldparams.RootLength]byte,
+	columns []uint64,
+) (map[uint64]bool, error) {
+	// distribute samples to peer
+	peerToColumns, err := d.distributeSamplesToPeer(columns)
+	if err != nil {
+		log.WithError(err).Error("Failed to distribute samples to peers")
+		return nil, errors.Wrap(err, "failed distributing samples to peer")
+	}
 
-	// Unsubscribe from the state feed when the function returns.
-	defer stateSub.Unsubscribe()
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	res := make(map[uint64]bool)
+	sampleFromPeer := func(pid peer.ID, cols map[uint64]bool) {
+		defer wg.Done()
+		retrieved := d.sampleDataColumnsFromPeer(ctx, pid, root, cols)
 
-	// Retrieve the number of columns.
-	columnsCount := params.BeaconConfig().NumberOfColumns
+		mu.Lock()
+		for col := range retrieved {
+			res[col] = true
+		}
+		mu.Unlock()
+	}
 
-	// Retrieve all columns we custody.
-	custodyColumns, err := peerdas.CustodyColumns(s.cfg.p2p.NodeID(), custodySubnetsCount)
-	if err != nil {
-		log.WithError(err).Error("Failed to get custody columns")
-		return
+	// sample from peers in parallel
+	for pid, cols := range peerToColumns {
+		wg.Add(1)
+		go sampleFromPeer(pid, cols)
 	}
 
-	custodyColumnsCount := uint64(len(custodyColumns))
+	wg.Wait()
+	return res, nil
+}
+
+// distributeSamplesToPeer distributes samples to peers based on the columns they are responsible for.
+// Currently it randomizes peer selection for a column and did not take into account whole peer distribution balance. It could be improved if needed.
+func (d *dataColumnSampler1D) distributeSamplesToPeer(
+	columns []uint64,
+) (map[peer.ID]map[uint64]bool, error) {
+	dist := make(map[peer.ID]map[uint64]bool)
 
-	// Compute the number of columns to sample.
-	if custodyColumnsCount >= columnsCount/2 {
-		log.WithFields(logrus.Fields{
-			"custodyColumnsCount": custodyColumnsCount,
-			"columnsCount":        columnsCount,
-		}).Debug("The node custodies at least the half the data columns, no need to sample")
-		return
+	for _, col := range columns {
+		peers := d.peerFromColumn[col]
+		if len(peers) == 0 {
+			return nil, errors.Errorf("no peers responsible for column %d", col)
```

Comment:
Not having any peer custodying a given column is for sure annoying. However, failing the whole sample process for that may be a little bit too extreme.

A better solution would be simply to declare this column as missing, and then a new round of lossyDAS will be triggered, and, hopefully, successful.

Reply:
Make sense